lib/gnat/services/service_responder.ex

```diff
@@ -0,0 +1,319 @@
+  def validate_configuration(configuration) do
+    rules = [
+      {&valid_version?/1, configuration},
+      {&valid_name?/1, configuration},
+      {&valid_metadata?/1, Map.get(configuration, :metadata)}
+    ]
+    eprules = configuration.endpoints
+    |> Enum.map(fn ep ->
+      {&valid_endpoint?/1, ep}
+    end)
+
+    results = (rules ++ eprules) |> Enum.map(fn {pred, input} ->
+      apply(pred, [input])
+    end)
+    {_good, bad} = Enum.split_with(results, fn e -> e == :ok end)
+
+    if length(bad) == 0 do
+      :ok
+    else
+      {:error, bad |> Enum.map(fn {:error, m} -> m end) |> Enum.to_list()}
+    end
+  end
+
+
+  @impl GenServer
+  @spec init(map) :: {:ok, state()} | {:stop, String.t}
+  def init(settings) do
+    Process.flag(:trap_exit, true)
+
+    case validate_configuration(Map.get(settings, :service_config)) do
+      {:error, errors} ->
+        {:stop, "Invalid service configuration: #{Enum.join(errors, ",")}"}
+      _ ->
+        state = %{
+          config: settings.service_config,
+          subject_map: build_subject_map(settings.service_config.endpoints),
+          connection_pid: Map.get(settings, :connection_pid),
+          instance_id: :crypto.strong_rand_bytes(12) |> Base.encode64,
+          subscription: nil,
+          started: DateTime.to_iso8601(DateTime.utc_now())
+        }
+
+        {:ok, subscription} = case get_in(state, [:config, :queue_group]) do
+          nil -> Gnat.sub(state.connection_pid, self(), @subscription_subject)
+          queue_group -> Gnat.sub(state.connection_pid, self(), @subscription_subject, queue_group: queue_group)
+        end
+
+        {:ok, %{ state | subscription: subscription }}
+      end
+
+  end
+
+  @impl true
+  def handle_info({:msg, %{topic: topic, body: _body, reply_to: rt, gnat: gnat}}, state) do
+    tokens = String.split(topic, ".")
+    if length(tokens) >= 2 do
+      operation = Enum.at(tokens, 1)
+      tail = tokens |> Enum.slice(2, length(tokens)-2)
+      case operation do
+        @op_ping -> handle_ping(tail, state, rt, gnat)
+        @op_info -> handle_service_info(tail, state, rt, gnat)
+        @op_stats -> handle_stats(tail, state, rt, gnat)
+      end
+    end
+
+    {:noreply, state}
+  end
+
+
+  @impl GenServer
+  def terminate(:shutdown, state) do
+    Logger.info "#{__MODULE__} starting graceful shutdown"
+    Gnat.unsub(state.connection_pid, state.subscription)
+    Logger.info "#{__MODULE__} finished graceful shutdown"
+  end
+  def terminate(reason, _state) do
+    Logger.error "#{__MODULE__} unexpected shutdown #{inspect reason}"
+  end
+
+  @impl true
+  def handle_cast({:record_request, subject, elapsed_ns}, state) do
+    counters = case :ets.lookup(:endpoint_stats, subject) do
+      [{^subject, counters, _last_error}] ->
+        counters
+      [] ->
+        c = :counters.new(3, [:atomics])
+        :ets.insert(:endpoint_stats, {subject, c, nil})
```

Comment:
I'm also not sure why we were inserting ETS records for the requests and errors? It looks like we only ever provide stats from the counters. Maybe in an earlier revision you were using ETS to count the requests and errors?

Reply:
Oh, maybe it's related to the `last_error` part of the stats callback? I didn't notice that on my first read-through of the spec, but the `last_error` and `data` keys seem a bit strange...

Reply:
the second element in the tuple is the counter set that contains the request and error counts.

Reply:
The only reason it's in `ets` is to allow a lookup of the counter set by subject when a request comes in. This lets us pull out the counter reference, increment, and move on.